File: dashboard/static/voice-gallery.js
'use strict';
(() => {
  // Rebrand the view without rewriting any previously exported private HTML.
  document.title = '语音档案 · 人类线上关系可视化';
  const records = JSON.parse(document.getElementById('voice-data').textContent);
  const el = id => document.getElementById(id);
  const labels = {done: '已转文字', pending: '待转写', empty: '未识别到文字', error: '转写失败', missing: '缺少音频', ambiguous: '关联待确认'};
  let page = 0;
  const pageSize = 40;
  const textNode = (tag, text, className) => {
    const node = document.createElement(tag);
    node.textContent = text;
    if (className) node.className = className;
    return node;
  };
  function render(reset) {
    if (reset) page = 0;
    const query = el('voice-query').value.trim().normalize('NFKC').toLocaleLowerCase();
    const state = el('voice-state').value;
    const from = el('voice-from').value;
    const to = el('voice-to').value;
    const sort = el('voice-sort').value;
    const filtered = records.filter(row => (!state || row.state === state)
      && (!query || `${row.contact}\n${row.text}`.normalize('NFKC').toLocaleLowerCase().includes(query))
      && (!from || row.date.slice(0, 10) >= from)
      && (!to || (row.date && row.date.slice(0, 10) <= to)));
    filtered.sort((a, b) => sort === 'longest' ? b.duration - a.duration : sort === 'contact'
      ? a.contact.localeCompare(b.contact, 'zh-CN') || b.date.localeCompare(a.date)
      : sort === 'oldest' ? a.date.localeCompare(b.date) : b.date.localeCompare(a.date));
    const pages = Math.max(1, Math.ceil(filtered.length / pageSize));
    page = Math.max(0, Math.min(page, pages - 1));
    const list = el('voice-list');
    list.querySelectorAll('audio').forEach(audio => audio.pause());
    list.replaceChildren();
    filtered.slice(page * pageSize, (page + 1) * pageSize).forEach(row => {
      const card = document.createElement('article');
      card.className = 'voice-card';
      const heading = document.createElement('div');
      heading.className = 'card-heading';
      heading.append(textNode('h2', row.contact), textNode('span', labels[row.state] || '待处理', 'badge ' + (row.state === 'done' ? 'done' : '')));
      card.append(heading, textNode('p', `${row.date || '日期未知'}${row.duration ? ' · ' + row.duration + ' 秒' : ''}`, 'meta'));
      if (row.audio && /^audio\/[a-f0-9]{64}\.wav$/.test(row.audio)) {
        const audio = document.createElement('audio');
        audio.controls = true;
        audio.preload = 'none';
        audio.src = row.audio;
        audio.setAttribute('aria-label', '播放这条语音');
        audio.addEventListener('play', () => list.querySelectorAll('audio').forEach(other => {if (other !== audio) other.pause();}));
        card.append(audio);
      }
      card.append(textNode('p', row.text || (row.state === 'missing' ? '目前本机未找到对应音频；聊天中的语音记录仍然保留。' : row.state === 'empty' ? '可以播放原音核对。未识别到可用文字。' : '文字尚未生成。'), row.text ? 'transcript' : 'muted'));
      list.append(card);
    });
    if (!filtered.length) list.append(textNode('p', '没有符合条件的语音，请调整关键词或日期。', 'empty'));
    el('voice-count').textContent = `共 ${filtered.length.toLocaleString()} 条匹配记录`;
    el('voice-page').textContent = `${page + 1} / ${pages}`;
    el('voice-prev').disabled = page === 0;
    el('voice-next').disabled = page + 1 >= pages;
  }
  for (const id of ['voice-query', 'voice-state', 'voice-from', 'voice-to', 'voice-sort']) el(id).addEventListener('input', () => render(true));
  el('voice-prev').addEventListener('click', () => {page--; render(false);});
  el('voice-next').addEventListener('click', () => {page++; render(false);});
  render(true);
})();

File: scripts/voice_gallery.py
"""Render a local, authenticated voice library; never print transcript text."""
from __future__ import annotations

from datetime import datetime
import html
import json
from pathlib import Path
import re

from message_merge import _atomic_write


def write_gallery(repo, catalog, results, status):
    export = Path(repo) / 'data/exports/wechat-voice'
    export.mkdir(parents=True, exist_ok=True)
    labels = {}
    rows = []
    for row in catalog.execute('SELECT * FROM voice_messages ORDER BY timestamp DESC,id'):
        bundle = row['bundle_id']
        if bundle not in labels:
            # Names stay in this local HTML, never in status or agent output.
            path = Path(repo) / 'data/contacts' / bundle / 'messages.json'
            try:
                from dashboard.search import _safe_path
                if not _safe_path(path, Path(repo) / 'data/contacts'):
                    raise ValueError
                value = json.loads(path.read_text('utf-8-sig'))
                labels[bundle] = str(value.get('contact_display') or value.get('contact_name') or '未命名会话')[:200]
            except (OSError, ValueError, TypeError):
                labels[bundle] = '未命名会话'
        cached = results.execute('SELECT * FROM transcripts WHERE sha=?', (row['sha'],)).fetchone()
        state = row['state']
        text, duration = '', 0
        if state == 'archived':
            state = cached['state'] if cached else 'pending'
            if cached:
                text, duration = cached['text'], cached['duration']
        try:
            timestamp = datetime.fromtimestamp(row['timestamp']).strftime('%Y-%m-%d %H:%M:%S')
        except (ValueError, OSError, OverflowError):
            timestamp = ''
        sha = row['sha'] or ''
        playable = bool(re.fullmatch('[a-f0-9]{64}', sha) and (export / 'audio' / (sha + '.wav')).is_file())
        rows.append({'contact': labels[bundle], 'date': timestamp, 'state': state,
                     'text': text, 'duration': round(duration, 1),
                     'audio': 'audio/' + sha + '.wav' if playable else ''})
    # Media without a message association are preserved and visible separately.
    for item in catalog.execute("SELECT * FROM assets WHERE sha NOT IN (SELECT sha FROM voice_messages WHERE sha != '')"):
        cached = results.execute('SELECT * FROM transcripts WHERE sha=?', (item['sha'],)).fetchone()
        sha = item['sha']
        rows.append({'contact': '未关联消息的录音', 'date': '', 'state': cached['state'] if cached else 'pending',
                     'text': cached['text'] if cached else '', 'duration': round(cached['duration'], 1) if cached else 0,
                     'audio': 'audio/' + sha + '.wav' if re.fullmatch('[a-f0-9]{64}', sha) and (export / 'audio' / (sha + '.wav')).is_file() else ''})
    payload = json.dumps(rows, ensure_ascii=False).replace('&', '\\u0026').replace('<', '\\u003c').replace('>', '\\u003e')
    total = len(rows)
    completed = int(status.get('transcribed_audio', 0))
    document = f'''<!doctype html>
<html lang="zh-CN"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1">
<title>语音档案 · 人类线上关系可视化</title><link rel="stylesheet" href="/voice-gallery.css"><script src="/voice-gallery.js" defer></script></head>
<body><main><nav><a href="/#maintenance">← 返回本地档案</a><span>本机语音库</span></nav>
<header><p class="eyebrow">VOICE / 声音与文字，一起留下</p><h1>把语音，变成能查找的记录。</h1>
<p class="intro">保留录音原件，用本地模型生成转写。重要内容请播放原音核对，尤其是人名、金额、方言和嘈杂片段。</p>
<div class="metrics"><div><b>{total:,}</b><span>语音记录与独立录音</span></div><div><b>{completed:,}</b><span>已生成文字的独立录音</span></div><div><b>{int(status.get('playable_audio', 0)):,}</b><span>可播放的独立录音</span></div></div></header>
<section class="filters" aria-label="筛选语音"><label class="search"><span>搜索联系人或转写内容</span><input id="voice-query" type="search" placeholder="输入关键词" autocomplete="off"></label>
<label><span>识别状态</span><select id="voice-state"><option value="">所有状态</option><option value="done">已转文字</option><option value="pending">待转写</option><option value="empty">未识别到文字</option><option value="error">转写失败</option><option value="missing">缺少音频</option><option value="ambiguous">关联待确认</option></select></label>
<label><span>开始日期</span><input id="voice-from" type="date"></label><label><span>结束日期</span><input id="voice-to" type="date"></label>
<label><span>排列顺序</span><select id="voice-sort"><option value="newest">最新在前</option><option value="oldest">最早在前</option><option value="longest">时长从长到短</option><option value="contact">联系人名称</option></select></label></section>
<div class="list-heading"><p id="voice-count" role="status"></p><p>录音仅在点击播放后加载</p></div>
<section id="voice-list" class="voice-list" aria-label="语音记录"></section>
<div class="pager"><button id="voice-prev" type="button">上一页</button><span id="voice-page"></span><button id="voice-next" type="button">下一页</button></div>
<footer>自动转写可能有误 · 搜索与转写在本机完成 · 云端分析需另外选择范围并勾选包含语音转写</footer>
<script id="voice-data" type="application/json">{payload}</script></main></body></html>'''
    target = export / 'index.html'
    content = document.encode('utf-8')
    if not target.exists() or target.read_bytes() != content:
        _atomic_write(target, content)
